### PR TITLE
Update `ComboCounter` components to use DI to attach data source

### PR DESCRIPTION
**Pull request by peppy**
This is the first of a few updates to how HUD components consume their data sources.

Starting with just one so I can make sure everyone is happy with the direction before converting the rest. I've tested this to work with skin editing and component toolbox as expected.

### CODE REVIEW

File: osu.Game/Screens/Play/HUDOverlay.cs [hunk 1]
```diff
@@ -276,13 +275,13 @@ protected virtual void BindDrawableRuleset(DrawableRuleset drawableRuleset)
...
+        protected SkinnableAccuracyCounter CreateAccuracyCounter() => new SkinnableAccuracyCounter();
 
-        protected virtual SkinnableScoreCounter CreateScoreCounter() => new SkinnableScoreCounter();
+        protected SkinnableScoreCounter CreateScoreCounter() => new SkinnableScoreCounter();
 
-        protected virtual SkinnableComboCounter CreateComboCounter() => new SkinnableComboCounter();
+        protected SkinnableComboCounter CreateComboCounter() => new SkinnableComboCounter();
 
-        protected virtual SkinnableHealthDisplay CreateHealthDisplay() => new SkinnableHealthDisplay();
+        protected SkinnableHealthDisplay CreateHealthDisplay() => new SkinnableHealthDisplay();
```
Comment (peppy) [inline osu.Game/Screens/Play/HUDOverlay.cs:284]: These will eventually be dismantled; just leaving in place for the time being (need to plan where these "defaults" will sit going forward).

File: osu.Game/Screens/Play/HUD/SkinnableComboCounter.cs [hunk 2]
```diff
@@ -1,29 +1,16 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
-using osu.Framework.Bindables;
 using osu.Game.Skinning;
 
 namespace osu.Game.Screens.Play.HUD
 {
-    public class SkinnableComboCounter : SkinnableDrawable, IComboCounter
+    public class SkinnableComboCounter : SkinnableDrawable
```
Comment (peppy) [inline osu.Game/Screens/Play/HUD/SkinnableComboCounter.cs:8]: I left this class in place as it feels cleander than shoving the `base()` spec inline at usages.

**Review by bdach — CHANGES_REQUESTED** [review]
No issues in principle with the approach here. Just a few tangential things to clean up maybe.

File: osu.Game/Screens/Play/HUD/SkinnableComboCounter.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Play/HUD/SkinnableComboCounter.cs:8]: Can remove if that's what we decide on, sure.

**Comment by frenzibyte** [comment]
I think this is in a pretty good state and direction, but I only have one slight issue with this, which is that `ScoreProcessor` is currently completely exposed to the skinnable elements in a mutable state, whereas it could be exposed as read-only instead via an `IScoreProcessor` interface.

**Comment by bdach** [comment]
Yeah I could get behind `IScoreProcessor`. Not sure whether it should be a thing in this PR or a separate one, but it'd be nice to have it in these usages.

**Comment by bdach** [comment]
Actually, looking at other usages of `ScoreProcessor` again, there are other components (`HitErrorDisplay`, `BreakTracker`, `MultiplayerGameplayLeaderboard`) that also receive a direct mutable reference to a `ScoreProcessor` without needing it to be mutable. So if this PR was to add an encapsulating interface, all of those should be using it too, which in my mind makes this a larger code quality change, out of scope of this one. So while I don't disagree with that direction I'm going to go ahead and approve this.

**Review by bdach — APPROVED** [review]

**Comment by frenzibyte** [comment]
Well, those components are explicitly getting it from constructor (`HitErrorDisplay` is a special case since that reference already has that removed in #12708), but yeah I don't see any problem in merging this and considering that a code-quality issue to be resolved in a separate PR. (after all the PRs in the chain have been merged)

**Review by frenzibyte — APPROVED** [review]

**Pull request merged**